kqueen/auth/ldap.py

```diff
@@ -22,31 +22,62 @@ def __init__(self, *args, **kwargs):
         self.connection = ldap.initialize(self.uri)
 
     @staticmethod
-    def _email_to_dn(email):
-        """This function reads email and converts it to LDAP dn
+    def _email_to_dn_dict(email):
+        """This function reads email and converts it to LDAP dn and dc's
 
         Args:
             email (str): e-mail address
 
         Returns:
-            dn (str): LDAP dn, like 'cn=admin,dc=example,dc=org'
+            dn (dict): LDAP dn and dc's, like {'cn': 'cn=admin', 'dcs': 'dc=example,dc=org'}
         """
 
-        segments = []
+        segments = {}
         if '@' in email:
             cn, dcs = email.split('@')
         else:
             cn = email
             dcs = ''
 
         if cn:
-            segments.append('cn={}'.format(cn))
+            segments['cn'] = cn
 
+        dc_list = []
         if '.' in dcs:
             for s in dcs.split('.'):
-                segments.append('dc={}'.format(s))
+                dc_list.append('dc={}'.format(s))
 
-        return ','.join(segments)
+        segments['dcs'] = ','.join(dc_list)
+        logger.debug('RegExp for search valid full-DN for Ldap user: {}'.format(segments))
+
+        return segments
+
+    def _get_matched_dn(self, username):
+        dn_segments = self._email_to_dn_dict(username)
+
+        if self._bind(self.kqueen_username, self.kqueen_password):
+            self.connection = ldap.initialize(self.uri)
+            self.connection.simple_bind_s(self.kqueen_username, self.kqueen_password)
+            self.connection.protocol_version = ldap.VERSION3
+
+            baseDN = dn_segments['dcs']
+            searchScope = ldap.SCOPE_SUBTREE
+            retrieveAttributes = ['dn']
+            searchFilter = "(cn={})".format(dn_segments['cn'])
+            ldap_result_id = self.connection.search_s(baseDN, searchScope, searchFilter, retrieveAttributes)
+
+            matched_dn = []
+            for key, value in dict(ldap_result_id).items():
+                matched_dn.append(key)
+
+            logger.debug('All matched to RegExp DN keys: {}'.format(matched_dn))
+            self.connection.unbind()
+            return matched_dn
+
+        else:
+            logger.debug('Failed to get matched to RegExp {} DN due connection problems'.format(matched_dn))
```

Comment:
looks like error, I guess we need to raise an exception here or at list set error log level. Is it handled above?

Reply:
it will handled by _bind() method exceptions